49 add support for h5md files

JFRudzinski opened the pull request:
@ladinesa I am finally trying to get back to the development of this H5MD-schema-based parser of hdf5 files.

I was wondering if you could take a (brief) initial look at this code (relevant corresponding nomad-lab dev [HERE](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/compare/develop...1547-add-support-for-h5md-files)) and also the initial docs that I have been working on [H5MD docs](https://fairmat-nfdi.github.io/AreaC-DOC/custom_schemas/h5md/h5md_overview/) just to get an initial sense of what I am trying to do. 

From there I think it would be nice to meet with you to discuss further before moving forward. I spoke with Markus about this a bit at the retreat, and we discussed the possibility of this parser being folded into a more general hdf5 parser. In either case, I realize that there are many things that I have done here that are not ideal, since I was more concerned with using this as a platform to define the schema itself. Additionally, with upcoming changes, I would like to discuss the most efficient way to move forward in a timely fashion.

JFRudzinski committed:
created h5md parser shell

JFRudzinski committed:
basic parser structure / code

JFRudzinski committed:
parse system and method complete, parse calculation in progress

JFRudzinski committed:
fixed connection to nomad and debugging parse_method

JFRudzinski committed:
finished with all basic parsing functions

JFRudzinski committed:
added bond_list to atoms and atomsgroup

JFRudzinski committed:
added bond_list to system[0].atoms in gromacs parser

JFRudzinski committed:
adjusted gromacs tests for interaction groupings

JFRudzinski committed:
reorganized interactions in lammps parse_method into groups, created bond list in system 0 atoms section, and moved the bond list creation to atomutils

JFRudzinski committed:
started tests for h5md

JFRudzinski committed:
done with initial tests except workflow results

JFRudzinski committed:
added basic metainfo for h5md atoms_group parameters unknown

JFRudzinski committed:
added custom params to atoms_group

JFRudzinski committed:
fixing setting of attributes in parse_calc

JFRudzinski committed:
implemented custom calculations and atomsgroup params

JFRudzinski committed:
start link to H5MD-NOMAD docs

JFRudzinski committed:
saving changes

JFRudzinski committed:
starting to implement workflow results

JFRudzinski committed:
working on parse_workflow

JFRudzinski committed:
added basic parsing of ensemble properties

JFRudzinski committed:
comment

JFRudzinski committed:
finished basic workflow results parsing

ladinesa commented:
I will have a closer look at this but just some initial thoughts:

1. Maybe I would split the parser itself with the h5 file parser. This way, we can reuse the h5 file parser for another parser.
2. Have you done some benchmarkings? I am not sure about the performance of accessing datasets one at a time but I think  traversing the h5 group one time will be faster. In the latter case, it will be harder to separate the parser from the file parser but it is doable.
3. This looks like an md parser so it would be nice the also use the md parser class or at least the parser will be made ready to easily transition to using it.

JFRudzinski commented:
> I will have a closer look at this but just some initial thoughts:
> 
> 1. Maybe I would split the parser itself with the h5 file parser. This way, we can reuse the h5 file parser for another parser.
> 2. Have you done some benchmarkings? I am not sure about the performance of accessing datasets one at a time but I think  traversing the h5 group one time will be faster. In the latter case, it will be harder to separate the parser from the file parser but it is doable.
> 3. This looks like an md parser so it would be nice the also use the md parser class or at least the parser will be made ready to easily transition to using it.

1. Yes, I agree. Is there anything existing already? I implemented a couple things that are schema-specific within the parsing of the file, so we would need to figure out how to make this work generally.
2. I have not at all. I knew that I would have to probably rewrite a lot, but it was easier conceptually for me to start this way. I probably need a bit of guidance here. I did sort of try to access the group only once by defining class properties that place the information within each group into a dict, but maybe I am not understanding correctly how this works.
3. Yes, I will definitely replace what I can by the super md parser (I started this before that was written actually).

JFRudzinski committed:
added list to thermostat params

JFRudzinski committed:
added back MDParser class

JFRudzinski committed:
fixed rebasing issues in lammps parser (sec_system definition was accidentally deleted)

JFRudzinski committed:
basic separation of hdf5 and h5md parsers

JFRudzinski committed:
corrected new metadata names in Interactions

JFRudzinski committed:
removed some commented out older code

JFRudzinski committed:
recovered previous lammps parser from #72 after rebasing error

JFRudzinski committed:
recovered previous gromacs parser from #72 after rebasing error

JFRudzinski committed:
implemented super-parser for parse_calculation, not yet tested

JFRudzinski committed:
working on integrating MDParser to H5MD

JFRudzinski committed:
implemented MDParser up to parse_calculation() in H5MD parser

JFRudzinski committed:
finished with mdworkflow method + MDParser with H5MD

JFRudzinski committed:
integrated super parser for H5MD

JFRudzinski committed:
only access outer hdf5 groups once

JFRudzinski commented:
@ladinesa Could you take a look at my changes here whenever you have a chance (no huge rush necessary)? I have tried to address your comments above. I still have a decent amount of testing/adjusting to do, but it would be great to hear your opinion if this is moving in the right direction.

btw - I plan to follow your changes to Gromacs and Lammps, and migrate this to the plugin schema, after rebasing my NOMAD branch

ladinesa commented:
> @ladinesa Could you take a look at my changes here whenever you have a chance (no huge rush necessary)? I have tried to address your comments above. I still have a decent amount of testing/adjusting to do, but it would be great to hear your opinion if this is moving in the right direction.
> 
> btw - I plan to follow your changes to Gromacs and Lammps, and migrate this to the plugin schema, after rebasing my NOMAD branch

Hi @JFRudzinski I will probably have time to look tomorrow.('',)

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        self.init_parser()
+
+        if self._data_parser.filehdf5 is None:
+            return
```
put warning here

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        if self._data_parser.filehdf5 is None:
+            return
+
+        sec_run = self.archive.m_create(Run)
```
would be nice to rid or m_create. We will be using inheritances now so these m_create will not anymore work for generic sections like EntryData, ArchiveSection. do
```python
sec_run = Run()
self.archive.run.append(sec_run)
```
This is more cumbersome, but makes it future-proof

JFRudzinski replied:
ok, I will avoid m_create from now on, thanks. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+
+        sec_run = self.archive.m_create(Run)
+
+        group_h5md = self.h5md_groups.get('h5md')
```
handle case when group_h5md is missing

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                except Exception:
+                    return
+        value = source.attrs.get(attribute)
+        source = value[-1] if isinstance(value, list) else value
```
maybe use value or another variable here instead of source.

JFRudzinski replied:
sure. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+
+        return quantity
+
+    def decode_hdf5_bytes(self, dataset, default):
```
you think we can drop hdf5 from function names, here decode_bytes for example?

JFRudzinski replied:
sure: decode_bytes(), attr_getter(), getter()? Or do you have other suggestions (especially for the last one I guess)?

ladinesa replied:
for consistency maybe decode_bytes, get_attribute, get_group?_dataset?_value?

JFRudzinski replied:
ok great, I did `get_group_dataset()` since I don't think there is a difference between a dataset or individual value. I think this is what you meant to have them together, but let me know if I misinterpreted.

ladinesa replied:
i am more inclined with get_value since value is more generic may mean group or dataset. get_group_dataset may imply getting the group with the dataset but if doc explains it, no problem for me.

JFRudzinski replied:
got it. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+
+        self._data_parser.mainfile = self.filepath
+
+    def parse(self, filepath, archive, logger):
```
i suggest we start doing type declarations? we can also do it during the migration.

JFRudzinski replied:
I went through and added declarations. There are still some instances that I am not sure how to deal with, I can go through and point these out to you at the next review time.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            for section in section_segments:
+                try:
+                    value = source.get(section)
+                    source = value[-1] if isinstance(value, list) else value
```
use group? instead of source here

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            for section in section_segments:
+                try:
+                    value = source.get(section)
+                    source = value[-1] if isinstance(value, list) else value
```
don't we put index as argument or we only return -1?

JFRudzinski replied:
I actually can't figure out now why I did this. It is actually never being triggered in my example. In fact, it seems that h5py is always converting lists to ndarrays when stored in the hdf5 file (not sure if that is happening in the writing or the reading step). In any case, I guess I will simply remove this statement and see if the situation every arises again.

ladinesa replied:
if i remember it right i always access data from dataset via dataset[()]

JFRudzinski replied:
yeah that is true. That is happening below this loop in the hdf5_getter. Again, not sure now why I was doing this. It is now simply removed, I will resolve this.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+    def decode_hdf5_bytes(self, dataset, default):
+        if dataset is None:
+            return
+        elif type(dataset).__name__ == 'ndarray':
```
can we use isinstance(dataset.?, np.ndarray)?

JFRudzinski replied:
indeed, that's better. Also fixed for the bool instance below. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+
+        return quantity
+
+    def decode_hdf5_bytes(self, dataset, default):
```
i suggest, not pass default here, just return None if not found, then handle None -> default in the calling function.

JFRudzinski replied:
ok, thanks. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        return self.decode_hdf5_bytes(source, default)
+
+    def parse(self, quantity_key: str = None, **kwargs):
+        pass
```
we can also interface the hdf5_getters to the get function here by populating parse something like:
```python
def parse(key, **kwargs):
    source = kwargs.get('source', self.filehdf5)
    path = kwargs.get('path')
    attribute = kwars.get('attribute')
    if attribute is not None:
        value = self.hdf5_attr_getter(source, path, attribute)
    else:
        value = self.hdf5_getter(source, path)
    # i do not know how to construct the full path here relative to root 
    full_path = f'...{path}.{attribute}'
    self._results[full_path] = value
```
This way, we also cache results.

JFRudzinski replied:
Is there an example that you can point me to that does something like this? All the parser classes that I can find either don't explicitly define parse() or pass within this function. I don't really understand this caching or what `full_path` is referring to

ladinesa replied:
you can have a look at the xml_parser.py in nomad.parsing.file_parser. The idea is to save the datasets in self._results[path.to.dataset] so that the next time you get(path.to.dataset), it is not fetched from the h5 file.

JFRudzinski replied:
Now implemented the parse() function and caching

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        sec_model = sec_force_field.m_create(Model)
+
+        # get the atom parameters
+        n_atoms = self.n_atoms[0]  # TODO Extend to non-static n_atoms
```
maybe error handling here if not self.n_atoms?

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        for n in range(n_atoms):
+            sec_atom = sec_method.m_create(AtomParameters)
+            for key in self.atom_parameters.keys():
+                setattr(sec_atom, key, self.atom_parameters[key][n])
```
do not use setattr, use instead
```python
sec_atom.m_set(sec_atom.get_quantity_defintion(key), self.atom_parameters[key][n])

JFRudzinski replied:
Should I also do instead of m_create?:
            sec_atom = AtomParameters()
            sec_method.atom_parameters.append(sec_atom)

ladinesa replied:
yes

JFRudzinski replied:
Also, with m_set, I get the error:
```
self = AtomParameters(), name = 'get_quantity_attribute'

    def __getattr__(self, name):
        # The existence of __getattr__ will make mypy and pylint ignore 'missing' dynamic
        # attributes and functions and wrong types of those.
        # Ideally we have a plugin for both that add the correct type info
    
        if name == 'm_key':
            return self.get(name)
    
        if name in self.m_def.all_aliases:
            return getattr(self, self.m_def.all_aliases[name].name)
    
        if self.m_def.has_variable_names:
            m_definition: Definition = resolve_variadic_name(self.m_def.all_properties, name)
            if m_definition:
                if not isinstance(m_definition, Quantity) or not m_definition.use_full_storage:
                    return getattr(self, m_definition.name)
    
                m_storage: dict = self.__dict__.get(m_definition.name, None)
                if m_storage is None:
                    return None
    
                m_quantity = m_storage.get(name, None)
                if m_quantity is None:
                    return None
    
                if m_quantity.value is not None:
                    if m_quantity.unit is not None:
                        return units.Quantity(m_quantity.value, m_quantity.unit)
    
                    return m_quantity.value
    
        if name.startswith('a_'):
            annotation_name = name[2:]
            if annotation_name not in self.m_annotations:
                raise AttributeError(name)
    
            return self.m_get_annotations(annotation_name)
    
>       raise AttributeError(name)
E       AttributeError: get_quantity_attribute

../../../nomad/metainfo/metainfo.py:1308: AttributeError
```

It's failing for `key = "label"`, which is a quantity in AtomParameters(). Any idea what is wrong here?

ladinesa replied:
sorry, it should be m_get_quantity_defintion()

JFRudzinski replied:
ah, yes, that's right, thanks. I thought I had used this before, but I couldn't find an example. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                setattr(sec_atom, key, self.atom_parameters[key][n])
+
+        # Get the interactions
+        connectivity = self.h5md_groups['connectivity']
```
```python
self.h5md_groups.get('connectivity')

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+    def h5md_groups(self):
+        if self._h5md_groups is None:
+            if not self._data_parser.filehdf5:
+                return
```
return {}

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        }
+
+    @property
+    def h5md_groups(self):
```
Do you think it is worthwhile to save these separately? I think we can simply access the group directly from root

JFRudzinski replied:
Do you mean defining a separate property for each group? I thought to do that first, but I wasn't sure which was better.

ladinesa replied:
simply having 
```python
self._data_parser.hdf5_getter(self._data_parser.filehdf5, 'particles') for example, since you can not executing each 'group'  more than once so saving the groups does not really offer any benefit.

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            interaction_list = self._data_parser.hdf5_getter(connectivity, interaction_key)
+            if interaction_list is None:
+                continue
+            elif type(interaction_list) == h5py.Group:
```
use isinstance(interaction_list, h5py.Group)

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            if interaction_list is None:
+                continue
+            elif type(interaction_list) == h5py.Group:
+                self.logger.warning('Time-dependent ' + key + ' currently not supported. ' + key + ' list will not be stored')
```
remove variables in warning simply: 'Time-dependent key not currently supported, will not be stored.'

JFRudzinski replied:
yes, sorry, I did this prior to our previous discussion about variables within warnings, and did not see this. removed.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        # Get the interactions
+        connectivity = self.h5md_groups['connectivity']
+        if not connectivity:
+            return
```
do we really return here? is connectivity also required futher down e.g.force_calculation_parameters

JFRudzinski replied:
yes, thank you. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        for key, val in force_calculation_parameters.items():
+            if not isinstance(val, dict):
+                key = self.check_metainfo_for_key_and_Enum(ForceCalculations, key, val)
+                setattr(sec_force_calculations, key, val)
```
use m_set

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            if particles_subgroup:
+                self.get_atomsgroup_fromh5md(sec_atomsgroup, particles_subgroup)
+
+    def check_metainfo_for_key_and_Enum(self, metainfo_class, key, val):
```
~~I think this is not necessary, base class will include x_ quantities if extend_base_section=True. ~~ not i get what you are trying to do here. but I would not access these private variables. simply do
```python
h5key = f'x_h5md_{key}'
key = key if hasattr(metainfo_class, key) else h5key if hasattr(metainfo_class, h5key) else None
quantity = getattr(key)
....
```
but I think you do not need to define the function. in case you decide to define the function: name it perhaps resolve_key

JFRudzinski replied:
ok, yeah, I was definitely not approaching this correctly. I think this part was outdated with how I am now approaching the non-normalized metadata. 

I now added the appropriate definitions in the h5md metainfo file. I still use a function to check the key/val pair (I need to cover the case that the key is valid but value does not match the MEnum list), but you can just see my changes in the next round of reviews.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                if key == 'neighbor_searching':
+                    for neigh_key, neigh_val in val.items():
+                        neigh_key = self.check_metainfo_for_key_and_Enum(NeighborSearching, neigh_key, neigh_val)
+                        setattr(sec_neighbor_searching, neigh_key, neigh_val)
```
use m_set

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                        neigh_key = self.check_metainfo_for_key_and_Enum(NeighborSearching, neigh_key, neigh_val)
+                        setattr(sec_neighbor_searching, neigh_key, neigh_val)
+                else:
+                    self.logger.warning(key + 'is not a valid force calculations section. Corresponding parameters will not be stored.')
```
remove varibale from msg

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            parameters_group = self.h5md_groups.get('parameters')
+            if parameters_group is None:
+                return
+            self._parameter_info = {
```
i think you return this if parameters_group is None, otherwise you will always do this whenever you access paramter_info.

JFRudzinski replied:
Can I just return {}? I can't populate these subsections in any case if there is no parameter info

JFRudzinski replied:
nm, done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                'workflow': {}
+            }
+
+            def get_parameters_recursive(parameter_group):
```
simply get_parameters

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            def get_parameters_recursive(parameter_group):
+                param_dict = {}
+                for key, val in parameter_group.items():
+                    if type(val) == h5py.Group:
```
insinstance

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                        param_dict[key] = get_parameters_recursive(val)
+                    else:
+                        param_dict[key] = self._data_parser.hdf5_getter(parameter_group, key)
+                        # if type(param_dict[key]) == str:
```
remove comment

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                                param_dict[key] = param_dict[key].upper()  # TODO change enums to lower case and adjust Gromacs and Lammps code accordingly
+                            else:
+                                param_dict[key] = param_dict[key].lower()
+                        elif 'int' in type(param_dict[key]).__name__:
```
isinstance(value, (int, np.int, np.int32))?

JFRudzinski replied:
indeed, thank you. well I added np.int64 instead of np.int. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        if workflow_parameters is None:
+            return
+
+        def get_workflow_results(property_type_dict, observables, workflow_results):
```
i suggest you make this a class method.. do not pass workflow_results as argument, simply return property_dict and update workflow_results outside the function. then it becomes get_workflow_property.

ladinesa replied:
be verbose with the arguments:
```python
def get_workflow_property(observables, property_key=None, property_value_key=None....)
```
then you call it
get_workflow_property(ensemble_average_observables, **ensemble_property_dict)

JFRudzinski replied:
I think I at least mostly executed what you wanted here. You probably should recheck in the next review round.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            'property_type_key': 'ensemble_properties',
+            'property_type_value_key': 'ensemble_property_values',
+            'properties_known': {'radial_distribution_functions': 'radial_distribution_function_values'},
+            'property_keys_list': EnsembleProperty.__dict__.keys(),
```
use EnsembleProperty.m_def.all_quantities.keys(), same for Correlation

JFRudzinski replied:
thanks. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+
+        system_info = self.system_info.get('system')
+        if not system_info:
+            self.logger.error('No particle information found in H5MD file.')
```
should we return here?

JFRudzinski replied:
yes, of course, thank you. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            positions_value = self.h5md_positions_value_all
+            if not particles_group:
+                return
+            self._system_info = {'system': {}, 'calculation': {}}
```
i think you should return dict(system={}...) if particles_group is None otherwise you call this everytime you access system_info

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+
+            if positions_value is None:  # For now we require that positions are present in the H5MD file to store other particle attributes
+                self.logger.warning('No positions available in H5MD file. Other particle attributes will not be stored')
+                return
```
also here return the empty dict

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                else:
+                    continue
+
+                if type(self._system_info[sec_key][key]) == h5py.Group:
```
isinstance

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                    if self._system_info[sec_key][key] is None:
+                        continue
+                    elif len(self._system_info[sec_key][key]) != n_frames:  # TODO Should really check that the stored times for these quantities are exact, not just the same length
+                        self.logger.warning('Distinct trajectory lengths of particle attributes not supported. ' + key + ' values will not be stored.')
```
remove variable

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                else:
+                    continue
+
+                if type(self._system_info[sec_key][box_key]) == h5py.Group:
```
isinstance

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                    if self._system_info[sec_key][box_key] is None:
+                        continue
+                    elif len(self._system_info[sec_key][box_key]) != n_frames:  # TODO Should really check that the stored times for these quantities are exact, not just the same length
+                        self.logger.warning('Distinct trajectory lengths of box vectors and positions is not supported. ' + box_key + ' values will not be stored.')
```
remove variable

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        if self._observable_info is None:
+            observables_group = self.h5md_groups.get('observables')
+            if observables_group is None:
+                return
```
return the empty dict you have in l328

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                'correlation_function': {}
+            }
+
+            def get_observable_paths(observable_group, current_path, paths):
```
do not pass paths but instead extend the paths outside the function. in l340 you have
```python
paths.extend(get_observable_paths(observable, f'{current_path}{path}'))

JFRudzinski replied:
sorry, I can't quite figure this one out. I don't understand where paths should be defined then, since I am calling the function recursively? Are you saying to completely remove paths as a function argument? Or just somehow not pass it when I call it within the function?

ladinesa replied:
you can initialize paths = [] in ll330 and then do paths.extend(get_observable_paths(observable, current_path + path)) in l335

JFRudzinski replied:
aha, ok thank you. I guess it was just intuitive to me for some reason what would happen with extend...thanks! done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                observable = self._data_parser.hdf5_getter(observables_group, path)
+                observable_type = self._data_parser.hdf5_getter(observables_group, path).attrs.get('type')
+                observable_name = path.split('.')[0]
+                observable_label = '-'.join(path.split('.')[1:]) if len(path.split('.')) > 1 else ''
```
you do not need the if here simply have
```python
observable_label = '-'.join(path.split('.')[1:])

JFRudzinski replied:
That doesn't quite work for what I am trying to achieve. when `path=group.name` it returns `name` which is what I want. but if `path=group`, I actually want to return an empty string, but yours returns `group`

ladinesa replied:
no i think i will work, index should be 1:

JFRudzinski replied:
ok, you are right, not sure what happened the first time :upside_down_face: thanks! done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                self._observable_info[observable_type][observable_name][observable_label] = {}
+                for key in observable.keys():
+                    observable_attribute = self._data_parser.hdf5_getter(observable, key)
+                    if type(observable_attribute) == h5py.Group:
```
isinstance

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                for key in observable.keys():
+                    observable_attribute = self._data_parser.hdf5_getter(observable, key)
+                    if type(observable_attribute) == h5py.Group:
+                        self.logger.warning('Group structures within individual observables not supported. ' + key + ' values will not be stored.')
```
remove variable

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+    def get_atomsgroup_fromh5md(self, nomad_sec, h5md_sec_particlesgroup):
+        for i_key, key in enumerate(h5md_sec_particlesgroup.keys()):
+            particles_group = {group_key: self._data_parser.hdf5_getter(h5md_sec_particlesgroup[key], group_key) for group_key in h5md_sec_particlesgroup[key].keys()}
+            sec_atomsgroup = nomad_sec.m_create(AtomsGroup)
```
do sec_atoms_group = AtomsGroup()

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                    self._observable_info[observable_type][observable_name][observable_label][key] = observable_attribute
+        return self._observable_info
+
+    def get_atomsgroup_fromh5md(self, nomad_sec, h5md_sec_particlesgroup):
```
this does seem to be a get function but parse_atomsgroup?

JFRudzinski replied:
indeed, done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            for particles_group_key in particles_group.keys():
+                unit = None
+                val = particles_group.get(particles_group_key)
+                has_units = hasattr(val, 'units') if val is not None else None
```
```python
units = val.units if hasattt(val, 'units') else None
if units is not None:
    val = val.magnitude

JFRudzinski replied:
indeed. such a weird code that I wrote! thanks. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+
+        n_frames = len(system_info.get('times', []))
+        self._system_time_map = {}
+        self._system_step_map = {}
```
i think you do not need this mapping, just provide
```python
self.trajectory_steps = list(range(n_frames))
self.thermodynamic_steps = [frame for frame in range(frame) if system_info.get('times, [None] * (frame + 1)) is not None]

JFRudzinski replied:
Each observable in `parse_calculation()` can have it's own individual list of trajectory times or steps. I think I need the mapping to make the complete, non-overlapping list of steps and then assign the system_refs properly, no? I do not see how I could do that simply with the lists that you defined.

ladinesa replied:
Thermo quantities are at different steps? Then this will not work unless you set self.thermodynamic_steps each each time you parse a quantity.

ladinesa replied:
No but I think this will work, thermodynamic_steps will simply be a list of all the steps for which a thermo property is calculated then set the property to None if not calculated at that step. Not sure if I read the situation clearl.y

ladinesa replied:
ok sorry, i think i understand now. the difficulty is that observable_info and system_info are saved as dict with key name, but if instead have time_step as key, we do not need to have an extra transformation and all the mapping. I do not like the the two nested loops in parse_calculation.

JFRudzinski replied:
So, I am still working on this. I need to still implement some of your suggestions, like the parsing function and removing observable_info as a property. In the mean time, I did re-implement the configurational part of observable_info as a dict by time, so that `parse_calculation` is much simpler.

However, I still have this nested loop to perform the reorganization:
```python
    def get_configurational_info_by_time(self):
        # reorganize the configurational observables by time step
        observable_dict = self.observable_info.get('configurational')
        configurational_dict_reorg = {}
        for observable_name, observable_name_dict in observable_dict.items():
            for observable_label, observable in observable_name_dict.items():
                times = observable.get('time')
                times = self.format_times(times) if times is not None else []
                for i_time, time in enumerate(times):
                    if not configurational_dict_reorg.get(time):
                        configurational_dict_reorg[time] = {}
                    if not configurational_dict_reorg[time].get(observable_name):
                        configurational_dict_reorg[time][observable_name] = {}
                    configurational_dict_reorg[time][observable_name][observable_label] = observable.get('value', [None] * (i_time + 1))[i_time]

        return configurational_dict_reorg
```

I will continue thinking about how to do it more efficiently, but I am open to suggestions :upside_down_face:

ladinesa replied:
I suggest, we simply have observable_dict expressed in terms of step rather than having to do these transformations?

JFRudzinski replied:
ok, I guess I do not know what you mean then. Here I am expressing the dictionary now by time_step. `observable_dict` is coming essentially directly from the hdf5 file. There the configurational observables **have** to be stored in an array. So, at some point I have to make this transformation to storing by time_step. I can also do this within the observable_info() function, but it will effectively be the same in terms of the nested loops. Am I missing something here?

JFRudzinski replied:
actually, I am working on redoing the parsing of system_info and observable_info, so I will get back to you on this in the next round

ladinesa replied:
> There the configurational observables **have** to be stored in an array

Do you really have to store this in an array or you can already implement the mapping here?

JFRudzinski replied:
I completely redid this part and also the parsing of the system_info. I think it is an improvement and at least closer to what you wanted. Let me know what you think though.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                        # TODO Still need to add custom values here.
+
+                    val = observable.get('value')
+                    if val is not None:
```
for l782-720, can we not define a common function for both value and bins? I do not quite like the repeated code.

JFRudzinski replied:
yes, of course. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+from .metainfo.h5md import ParamEntry, CalcEntry, Author
+from nomad.units import ureg
+
+MOL = 6.022140857e+23
```
can you please put this def in utils?

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+        return self._h5md_groups
+
+    @property
+    def h5md_particle_group_all(self):
```
for defs l174, l183, l191, I suggest you do not make them as property, make them simply as private variable e.g. self._h5md_particle_group_all which you calculate at the beginning of def parse.

JFRudzinski replied:
Ah ok. got it.

Just a general question about this parse() function: I guess this is automatically triggered, but when/how exactly does this happen? When a class is instantiated?

ladinesa replied:
there are a lot of layers with the addition of plugin (plugin-MatchingParser-MatchingParserInterface-Parser)..the parser class is only instantiated once and parse is executed everytime you parse a mainfile so all intitializations should be here. The parser class is istantiated in l411 of nomad.parsing.parser.py and parse is called in l423 which is triggered in l1203 in nomad.processing.data.py

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+    def atom_parameters(self):
+        if self._atom_parameters is None:
+            if not self.h5md_particle_group_all:
+                return
```
return self._atom_parameters = {}

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+            self._atom_parameters = {}
+            n_atoms = self.n_atoms[0]  # TODO Extend to non-static n_atoms
+            if n_atoms is None:
+                return
```
also here

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                    self._atom_parameters[key] = value
+                else:
+                    continue
+                if type(self._atom_parameters[key]) == h5py.Group:
```
isinstance

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                else:
+                    continue
+                if type(self._atom_parameters[key]) == h5py.Group:
+                    self.logger.warning('Time-dependent ' + key + ' currently not supported. Atom parameter values will not be stored.')
```
variable

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,787 @@
+                    self.logger.warning('Time-dependent ' + key + ' currently not supported. Atom parameter values will not be stored.')
+                    continue
+                elif len(self._atom_parameters[key]) != n_atoms:
+                    self.logger.warning('Inconsistent length of ' + key + ' . Atom parameter values will not be stored.')
```
variable

JFRudzinski replied:
done.

ladinesa commented on tests/test_h5md.py:
```diff
@@ -0,0 +1,190 @@
+    return H5MDParser()
+
+
+def test_md(parser):
```
does this test also traj sampling and stuff? Maybe good to also have test where some quantities are missing like system_info. Not sure if you can generate such a file.

JFRudzinski replied:
yeah, I agree. I would say that I need to generate a whole variety of tests to try to cover as many scenarios as possible, since the file format has so much flexibility. I start working on this..

ladinesa commented on tests/test_h5md.py:
```diff
@@ -0,0 +1,190 @@
+    sec_atom_params = sec_method[0].atom_parameters
+    assert len(sec_atom_params) == 31583
+    assert sec_atom_params[164].label == 'O'
+    assert sec_atom_params[164].mass.magnitude == approx(2.656767806475139e-26)
```
sorry if I will make this more difficult for you but we should compare the number that is parsed , so you would have mass.to(h5unit).magnitude. But this could be fixed later.

JFRudzinski replied:
actually, there is no inherent h5md unit. The user can specify the unit using the UnitRegistry-compatible strings.

I'm curious though why we need to compare in the "native" units when applicable (this is certainly not what the gromacs and lammps tests do currently)? If there was a mistake in the parsing we will certainly notice after conversion to a different unit, right?

ladinesa replied:
I think it is simply more convenient if we compare to the value that is 'printed on the mainfile'. Yes, this is not what is currently done in most parsers but we are trying to move in this direction. This is a remnant of the old infra where we do not support unit conversions.

JFRudzinski replied:
Do you mean here simply to convert the value within the test to the unit found in the original file and then enter the value from the file (instead of from the parsed result)?

ladinesa replied:
yes, do something like, quantity.to('orignal_unit').magnitude == approx(raw_value_in_file)

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser_stripped.py:
```diff
@@ -0,0 +1,275 @@
+#
```
not sure what this file is

JFRudzinski replied:
removed.

ladinesa commented on atomisticparsers/h5md/nomad_plugin.yaml:
```diff
@@ -0,0 +1,18 @@
+code_category: Atomistic code
+code_homepage: http://www.??.org/
```
?

JFRudzinski replied:
I commented these urls out for now, and put a TODO to check if this is the correct category. The most relevant link here will end up being the docs page of the eventual "h5md-nomad" plugin that this parser turns into...but I am not sure how we deal with these metadata in general atm

ladinesa commented:
I think I have covered most parts, we can do another round if you like after you implemented my suggestions. This is already in good form.

ladinesa commented:
> @ladinesa I am finally trying to get back to the development of this H5MD-schema-based parser of hdf5 files.
> 
> I was wondering if you could take a (brief) initial look at this code (relevant corresponding nomad-lab dev [HERE](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/compare/develop...1547-add-support-for-h5md-files)) and also the initial docs that I have been working on [H5MD docs](https://fairmat-nfdi.github.io/AreaC-DOC/custom_schemas/h5md/h5md_overview/) just to get an initial sense of what I am trying to do.
> 
> From there I think it would be nice to meet with you to discuss further before moving forward. I spoke with Markus about this a bit at the retreat, and we discussed the possibility of this parser being folded into a more general hdf5 parser. In either case, I realize that there are many things that I have done here that are not ideal, since I was more concerned with using this as a platform to define the schema itself. Additionally, with upcoming changes, I would like to discuss the most efficient way to move forward in a timely fashion.

for the nomad part, i was not able to find the mr. you have no open mr. please create drafts for these so I can have a look, thanks

JFRudzinski committed:
partial fixes from Alvin's review

JFRudzinski committed:
added metainfo for x_h5md_parameters for force calculations, and implemented check of valid key-val pair

JFRudzinski committed:
more review fixes

JFRudzinski committed:
shifted get_workflow_results() to class method

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,784 @@
+                    if key == 'forces':
+                        data[key] = dict(total=dict(value=val[system_index]))
+                    else:
+                        if key in BaseCalculation.__dict__.keys():
```
hasattr(BaseCalculation, key)

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,784 @@
+                    if obs_index:
+                        val = observable.get('value', [None] * (obs_index + 1))[obs_index]
+                        if 'energ' in observable_type:  # TODO check for energies or energy when matching name
+                            if key in Energy.__dict__.keys():
```
hasattr(Energy, key)

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,784 @@
+                            else:
+                                key = map_key
+
+                            if key in BaseCalculation.__dict__.keys():
```
same here.

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,784 @@
+        return self._system_info
+
+    @property
+    def observable_info(self):
```
looking at how observable_info is used in the code, I think we do not need to make this a property. you can simply have a function
```python
def get_observable_info(self, observable_type):
```
I would like to avoing storing duplicate data.

JFRudzinski replied:
yeah, it's not super clear to me when I should be using @property, I need to review this. Does this apply to the other `xxx_info()` functions as well?

ladinesa replied:
I will try to look at the others also. but the rule is, we use property if a variable is dependent on some state and needs  to be updated. in this case, it is constant and we should simply use a private variable or a funtion to calc it.

JFRudzinski replied:
that makes sense, thanks. I think probably none of these `xxx_info()` need to be properties then, then are simply parsing the hdf5 groups.

JFRudzinski replied:
converted all these `xxx_info()` to parsing functions now (no longer properties).

JFRudzinski committed:
another round of review fixes

JFRudzinski committed:
reorg observable_info by time

JFRudzinski committed:
direct definitions of h5md groups

JFRudzinski committed:
made several of the former class properties simply defined as private variables within parse()

JFRudzinski committed:
changed xxx_info() to parse functions (not properties)

JFRudzinski committed:
completely redid parsing of system_info, system, observable_info, observable

JFRudzinski commented:
@ladinesa Thanks a lot for all your work on this initial review, I found it **very** helpful.

The main things that I have not yet addressed are 1. implementation of the parse() function within the hdf5 file parser and 2. the creation of a slew of tests for various situations.

I'm sure there are many other things to improve. I will have another critical look early next year. FYI - I probably won't have much more time to work on this before my vacation (starting Thu), but it will be towards the top of my priority when I return on Jan 2.

ladinesa commented:
> @ladinesa Thanks a lot for all your work on this initial review, I found it **very** helpful.
> 
> The main things that I have not yet addressed are 1. implementation of the parse() function within the hdf5 file parser and 2. the creation of a slew of tests for various situations.
> 
> I'm sure there are many other things to improve. I will have another critical look early next year. FYI - I probably won't have much more time to work on this before my vacation (starting Thu), but it will be towards the top of my priority when I return on Jan 2.

We can continue again next year. But very good progress. I will also be back Jan 10.

JFRudzinski committed:
implemented initial functionality of the parse function within the hdf5 file parser

JFRudzinski committed:
full implementation of parse() with old code commented

JFRudzinski committed:
removed comments and clean

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,692 @@
+        isattr = kwargs.get('isattr', False)
+        value = None
+        if isattr:
+            attr_path = path.split('.')
```
attr_path, attribute = path.rsplit('.', 1)

JFRudzinski replied:
thanks. done.

JFRudzinski committed:
test in raw file units

JFRudzinski commented:
@ladinesa I have now addressed all your review comments. I still need to generate more tests, but that will take me some time. In the mean time, I was wondering if we should already think about transferring this to a plugin, or is it better to wait until you get further with some of the other parsers? In any case, I think another review of the code would be good, if you have time at some point. Thanks!

ladinesa commented:
> @ladinesa I have now addressed all your review comments. I still need to generate more tests, but that will take me some time. In the mean time, I was wondering if we should already think about transferring this to a plugin, or is it better to wait until you get further with some of the other parsers? In any case, I think another review of the code would be good, if you have time at some point. Thanks!

My plan for the development of the individual parser plugins is to already include the new data schema. That is why I hold off the progress with the other parsers. I suggest that we keep it here for the moment.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            'dimension': 'dimension'
+        }
+
+    def format_times(self, times):
```
i do not see any references to this.

JFRudzinski replied:
yes, thanks, this is an artifact from something we changed. removed.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+                if value is None or attr_steps is None:
+                    self.logger.warning('Missing values or steps in particle attributes. These attributes will not be stored.')
+                    return None
+                elif attr_steps.sort() != steps.sort() if attr_steps is not None else False:
```
.sort() will return None, do you want to compare the sorted list? if so, this would not work and will always return true. also, maybe simplify the condition by not using else if, try instead attr_steps and sorted(attr_step) != sorted(steps)

JFRudzinski replied:
Thank you! I did not realize this. `elif sorted(attr_steps) != sorted(steps) if attr_steps is not None else False:` works right?

Actually, I think I only need `elif sorted(attr_steps) != sorted(steps)` since I already check `attr_steps is None` in the first if statement.

ladinesa replied:
> Thank you! I did not realize this. `elif sorted(attr_steps) != sorted(steps) if attr_steps is not None else False:` works right?
> 
> Actually, I think I only need `elif sorted(attr_steps) != sorted(steps)` since I already check `attr_steps is None` in the first if statement.

o yes you are right.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            else:
+                continue
+            if isinstance(self._atom_parameters[key], h5py.Group):
+                self.logger.warning('Time-dependent atom parameters currently not supported. Atom parameter values will not be stored.')
```
Please use line breaks for really long comments

JFRudzinski replied:
changed throughout.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            if value is None:
+                return value
+            if isinstance(value, h5py.Group):
+                path_value = f'{path}.value' if path else 'value'
```
for cases like this, i would simply not define path_value and put the expression directly in .get(...)

JFRudzinski replied:
got it. done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+        # get the box quantities
+        box = self._data_parser.get(f'{self._path_group_particles_all}.box')
+        if box is not None:
+            box_attributes = {'dimension': 'system', 'periodic': 'system'}
```
will we have a case where the section key is not system? if not why use dict? is it the case that you want it to be consistent with l234

JFRudzinski replied:
I think probably not. Yes, I was just trying to be consistent and leave it general for the future, but I think now it's not necessary. I changed box_attributes to a list now.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+                path = f'{self._path_group_particles_all}.box.{self._nomad_to_box_group_map[box_key]}'
+                value = self._data_parser.get(path, isattr=True)
+                values_dict[sec_key][box_key] = [value] * n_frames if value is not None else None
+            box_keys = {'lattice_vectors': 'system'}
```
also here, and will we add more keys in box_keys?

JFRudzinski replied:
Here I even removed the loop.

ladinesa replied:
yes thanks.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+        def get_observable_paths(observable_group: Dict, current_path: str) -> List:
+            paths = []
+            for obs_key in observable_group.keys():
+                path = obs_key + '.'
```
please change to f-formatting for all instances

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            full_path = f'observables.{path}'
+            observable = self._data_parser.get_value(observables_group, path)
+            observable_type = self._data_parser.get_value(observables_group, path).attrs.get('type')
+            observable_name = path.split('.')[0]
```
observable_name, observable_label = path.split('.', 1)

JFRudzinski replied:
changed to `observable_name, observable_label = path.split('.', 1) if len(path.split('.')) > 1 else [path, '']` for the case that path has no '.' e.g., path = 'test', which would throw an error for the suggested code

ladinesa replied:
ah yes sorry, then please do not do one line because this is not also pretty we split twice. do instead:
```python
name_label = path.split('.', 1)
observable_name = path[0]
observable_label = path[1] if len(path) > 1 else ''

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+                if steps is None:
+                    self.logger.warning('Missing step information in some observables. These will not be stored.')
+                    continue
+                thermodynamics_steps = list(set(steps) | set(thermodynamics_steps))
```
i really like using more descriptive operators, for this maybe set(list(steps) + thermodynamic_steps)

JFRudzinski replied:
I would then have to do `thermodynamics_steps = set(list(steps) + list(thermodynamics_steps))` since `thermodynamics_steps` is a set in each case. Do you still prefer that?

ladinesa replied:
it is actually more readable, but your call, | is more compact

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+
+            self.thermodynamics_steps = thermodynamics_steps
+
+    def parse_atomsgroup(self, nomad_sec, h5md_sec_particlesgroup: Group, path_particlesgroup: str):
```
nomad_sec type is either System or AtomsGroup.

JFRudzinski replied:
Thanks, done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+    def parse_atomsgroup(self, nomad_sec, h5md_sec_particlesgroup: Group, path_particlesgroup: str):
+        for i_key, key in enumerate(h5md_sec_particlesgroup.keys()):
+            path_particlesgroup_key = f'{path_particlesgroup}.{key}'
+            particles_group = {group_key: self._data_parser.get(f'{path_particlesgroup_key}.{group_key}') for group_key in h5md_sec_particlesgroup[key].keys()}
```
line break here

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            if particles_subgroup:
+                self.parse_atomsgroup(sec_atomsgroup, particles_subgroup, f'{path_particlesgroup_key}.particles_group')
+
+    def is_valid_key_val(self, metainfo_class, key: str, val) -> bool:
```
metainfo_class type is MSection

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+    def is_valid_key_val(self, metainfo_class, key: str, val) -> bool:
+        if hasattr(metainfo_class, key):
+            quant_type = getattr(metainfo_class, key).get('type')
+            is_MEnum = isinstance(quant_type, MEnum) if quant_type else False
```
use snake case. is_menum = quant_type and isinstance(quant_type, MEnum)

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+        if hasattr(metainfo_class, key):
+            quant_type = getattr(metainfo_class, key).get('type')
+            is_MEnum = isinstance(quant_type, MEnum) if quant_type else False
+            return False if is_MEnum and val not in quant_type._list else True
```
return not is_menum and val not in quan_type._list

JFRudzinski replied:
This doesn't quite work cause I can only access `quant_type._list` if `is_menum` is True. In any case, I just simplified it in this way:
``` python
def is_valid_key_val(self, metainfo_class: MSection, key: str, val) -> bool:
    if hasattr(metainfo_class, key):
        quant_type = getattr(metainfo_class, key).get('type')
        is_menum = isinstance(quant_type, MEnum) if quant_type else False
        if is_menum:
            return val in quant_type._list
        else:
            return True
    else:
        return False
```
But if you have another suggestion let me know

ladinesa replied:
my bad, original code looks better.

JFRudzinski replied:
reverted to old code.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+                    param_dict[key] = self._data_parser.get(path_key)
+                    if isinstance(param_dict[key], str):
+                        if key == 'thermodynamic_ensemble':
+                            param_dict[key] = param_dict[key].upper()  # TODO change enums to lower case and adjust Gromacs and Lammps code accordingly
```
one line if else

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+        force_calculations_group = self._data_parser.get(path_force_calculations)
+        if force_calculations_group is not None:
+            self._parameter_info['force_calculations'] = get_parameters(force_calculations_group, path_force_calculations)
+        path_workflow = 'parameters.workflow'
```
put directly in .get

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+    def parse_calculation(self):
+        sec_run = self.archive.run[-1]
+        calculation_info = self._observable_info.get('configurational')
+        system_info = self._system_info.get('calculation')  # note: it is currently ensured in parse_system() that these have the same length as the system_map
```
put after l390

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+                'x_h5md_custom_calculations': [],
+                'x_h5md_energy_contributions': []
+            }
+            data['time'] = calculation_info.get('step', {}).get('time') # * self._time_unit
```
remove dangling comment

JFRudzinski replied:
thanks, done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            if not data['time']:
+                data['time'] = system_info.get('step', {}).get('time')
+
+            if system_info.get(step):
```
remove if, for key, val in system_info.get(step, {})

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+                                val = val.magnitude
+                            data_h5md['x_h5md_custom_calculations'].append(CalcEntry(kind=key, value=val, unit=unit))
+
+            if calculation_info.get(step):
```
for key, val in calculation_info.get(step, {})

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+                    key_split = key.split('-')
+                    observable_name = key_split[0]
+                    observable_label = key_split[1] if len(key_split) > 1 else key_split[0]
+                    if 'energ' in observable_name:  # TODO check for energies or energy when matching name
```
in case if name starts with energ I suggest using .startswith('energ') otherwise also fine.

JFRudzinski replied:
This is a bit too restrictive, e.g., things like `kinetic_energy` need to be identified

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            sec_calc = sec_run.calculation[-1]
+
+            if sec_calc.step != step:  # TODO check this comparison
+                sec_calc = sec_run.m_create(Calculation)
```
let's keep away from using m_create. More future proof to do 
sec_calc = Calculation()
sec_run.calculation.append(sec_calc)

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+
+        sec_method = Method()
+        self.archive.run[-1].method.append(sec_method)
+        sec_force_field = sec_method.m_create(ForceField)  # TODO @landinesa how do we remove m_create's here?
```
yes please avoid using m_create. The reason being that since we are moving in the direction of base classes, m_create would not work in case the parent class was used to define a sub_section.

JFRudzinski replied:
yes, I wanted to change this, but I cannot figure out the correct approach in this case:

```
        sec_force_field = ForceField()
        self.archive.run[-1].method[-1].force_field.append(sec_force_field)
```
AttributeError: 'NoneType' object has no attribute 'append'...but I don't understand because in my mind this should be the same as how we treat the run section earlier in the code??

ladinesa replied:
force_field is not a repeating section so you should do self.archive.run[-1].method[-1].force_field = sec_force_field

JFRudzinski replied:
ok, sorry, that was stupid. I figured this out:
```
  sec_method = Method()
  self.archive.run[-1].method.append(sec_method)
  sec_force_field = ForceField()
  sec_method.force_field = sec_force_field
  sec_model = Model()
  sec_force_field.model.append(sec_model)
```
Is this sufficient?

ladinesa replied:
it is really not obvious that repeating sections default to an empty list, otherwise None. I also just realized it recently. yes that is correct.

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+        # Get the force calculation parameters
+        force_calculation_parameters = self._parameter_info.get('force_calculations')
+        if force_calculation_parameters:
+            sec_force_calculations = sec_force_field.m_create(ForceCalculations)
```
also here.

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+                        units = val.units if hasattr(val, 'units') else None
+                        val = val.magnitude if units is not None else val
+                        sec_force_calculations.x_h5md_parameters.append(ParamEntry(kind=key, value=val, unit=units))
+                else:
```
the else here is not necessary use: elif key == 'neighbor_searching' else:

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+
+        workflow_parameters = self._parameter_info.get('workflow').get('molecular_dynamics')
+        # TODO should store parameters that do not match the enum vals as x_h5MD params,
+        # not sure how with MDParser??
```
can you please elaborate?

JFRudzinski replied:
Sorry, this is an outdated comment. removed.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+        }
+        workflow_results.update(
+            self.get_workflow_properties_dict(ensemble_average_observables, **ensemble_property_dict)
+            )
```
overindented )

JFRudzinski replied:
Like this:
```
        workflow_results.update(
            self.get_workflow_properties_dict(ensemble_average_observables, **ensemble_property_dict)
        )
```
??

ladinesa replied:
yes, not sure if the linter will actually pick it up but just to be safe hehe.

JFRudzinski replied:
yeah, the ruff extension for vscode auto-fixes a lot of spacing and such, but doesn't catch this for some reason.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            )
+        self.parse_md_workflow(dict(method=workflow_parameters, results=workflow_results))
+
+    def init_parser(self):
```
remove init_parser put all init in parse

JFRudzinski replied:
done.

ladinesa commented on atomisticparsers/h5md/parser.py:
```diff
@@ -0,0 +1,690 @@
+            return
+
+        # Define some useful paths within the hdf5 file
+        self._path_group_particles_all = 'particles.all'
```
are these not supposed to be in __init__

JFRudzinski replied:
I moved these up where the other variables are specified within parse(), not sure if this is what you meant.

ladinesa replied:
do these change? if not put them in 'def \__init__' like all the other constants.

JFRudzinski replied:
oh :facepalm: got it, done.

ladinesa commented:
I am satisfied now with the current state. you can merge this at will. thanks!

JFRudzinski committed:
most 2nd review changes

JFRudzinski committed:
more changes from review #2
